src/style-spec/function/expressions.js

```diff
@@ -0,0 +1,354 @@
+                    return {
+                        errors: [ 'Input/output pairs for "curve" expressions must be defined using literal numeric values (not computed expressions) for the input values.' ]
+                    };
+                }
+
+                if (stops.length && stops[stops.length - 1] >= input.value) {
+                    return {
+                        errors: [ 'Input/output pairs for "curve" expressions must be arranged with input values in strictly ascending order.' ]
+                    };
+                }
+
+                stops.push(input.value);
+                outputs.push(`() => ${output.js}`);
+            }
+
+            if (stops.length === 1) return {js: `${outputs[0]}`};
+
+            const interpolationOptions: Object = {
+                name: interpolation.name
+            };
+
+            if (interpolation.name === 'exponential') {
+                const baseExpr = interpolation.arguments[0];
+                if (!baseExpr.literal || typeof baseExpr.value !== 'number') {
+                    return {errors: ["Exponential interpolation base must be a literal number value."]};
+                }
+                interpolationOptions.base = baseExpr.value;
+            }
+
+            // TODO: investigate how this code is optimized in V8
+            return {js: `
+            (function () {
+                var input = ${args[1].js};
+                var stopInputs = [${stops.join(', ')}];
+                var stopOutputs = [${outputs.join(', ')}];
+                return this.evaluateCurve(${args[1].js}, stopInputs, stopOutputs, ${JSON.stringify(interpolationOptions)}, ${JSON.stringify(resultType)});
+            }.bind(this))()`};
+        }
+    },
+    'step': {
+        name: 'step',
+        type: lambda(InterpolationType),
+        compile: () => ({ js: 'void 0' })
+    },
+    'exponential': {
+        name: 'exponential',
+        type: lambda(InterpolationType, NumberType),
+        compile: () => ({ js: 'void 0' })
+    },
+    'linear': {
+        name: 'step',
+        type: lambda(InterpolationType),
+        compile: () => ({ js: 'void 0' })
+    }
+};
+
+module.exports = expressions;
+
+function defineMathConstant(name) {
+    const mathName = name.toUpperCase();
+    assert(typeof Math[mathName] === 'number');
+    return {
+        name: name,
+        type: lambda(NumberType),
+        compile: () => ({ js: `Math.${mathName}` })
+    };
+}
+
+function defineMathFunction(name: ExpressionName, arity: number, mathName?: string) {
+    const key:string = mathName || name;
+    assert(typeof Math[key] === 'function');
+    assert(arity > 0);
+    const args = [];
+    while (arity-- > 0) args.push(NumberType);
+    return {
+        name: name,
+        type: lambda(NumberType, ...args),
+        compile: (_, args) => ({ js: `Math.${key}(${args.map(a => a.js).join(', ')})` })
+    };
+}
+
+function defineBinaryMathOp(name, isAssociative) {
```

Comment:
Ah, whoops!  This was intended for allowing  `+`, `&&`, etc. to take > 2 args as a convenience.